src/components/createElementComponent.tsx

```diff
@@ -55,170 +53,106 @@ const createElementComponent = (
-        // just as they could listen for the `checkout` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on(
-          'checkout',
-          (event: stripeJs.StripeCartElementPayloadEvent) => {
-            if (type === 'cart' && setCartState) {
-              // we know that elements.on event must be of type StripeCartPayloadEvent if type is 'cart'
-              // we need to cast because typescript is not able to infer which overloaded method is used based off param type
-              setCartState(event);
-            }
-            callOnCheckout(event);
+    const checkoutCallback =
+      type === 'cart'
+        ? (event: stripeJs.StripeCartElementPayloadEvent) => {
+            setCartState(event);
+            onCheckout && onCheckout(event);
           }
-        );
-
-        // Users can pass an onLineItemClick prop on any Element component
-        // just as they could listen for the `lineitemclick` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on('lineitemclick', callOnLineItemClick);
-
-        // Users can pass an onConfirm prop on any Element component
-        // just as they could listen for the `confirm` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on('confirm', callOnConfirm);
+        : onCheckout;
 
-        // Users can pass an onCancel prop on any Element component
-        // just as they could listen for the `cancel` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on('cancel', callOnCancel);
+    useAttachEvent(element, 'checkout', checkoutCallback);
 
-        // Users can pass an onShippingAddressChange prop on any Element component
-        // just as they could listen for the `shippingaddresschange` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on(
-          'shippingaddresschange',
-          callOnShippingAddressChange
-        );
-
-        // Users can pass an onShippingRateChange prop on any Element component
-        // just as they could listen for the `shippingratechange` event on any Element,
-        // but only certain Elements will trigger the event.
-        (element as any).on('shippingratechange', callOnShippingRateChange);
+    React.useLayoutEffect(() => {
+      if (element === null && elements && domNode.current !== null) {
+        const newElement = elements.create(type as any, options);
+        if (type === 'cart' && setCart) {
+          // we know that elements.create return value must be of type StripeCartElement if type is 'cart',
+          // we need to cast because typescript is not able to infer which overloaded method is used based off param type
+          setCart((newElement as unknown) as stripeJs.StripeCartElement);
+        }
+        setElement(newElement);
```

Comment:
This causes an extra render compared to before, but this seemed cleaner than modifying `useAttachEvent` to return some sort of `addEvent` function that is manually called here.